bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperBuilderImpl.java

```diff
@@ -0,0 +1,94 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client;
+
+import com.google.common.base.Preconditions;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.HashedWheelTimer;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import static org.apache.bookkeeper.client.BKException.Code.ZKException;
+import org.apache.bookkeeper.client.api.BookKeeper;
+import org.apache.bookkeeper.client.api.BookKeeperBuilder;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.feature.FeatureProvider;
+import org.apache.bookkeeper.net.DNSToSwitchMapping;
+import org.apache.bookkeeper.stats.NullStatsLogger;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooKeeper;
+
+/**
+ * Internal builder for {@link org.apache.bookkeeper.client.api.BookKeeper} client
+ */
+public class BookKeeperBuilderImpl implements BookKeeperBuilder {
+
+    private final ClientConfiguration conf;
+    private ZooKeeper zk = null;
+    private EventLoopGroup eventLoopGroup;
+    private StatsLogger statsLogger = NullStatsLogger.INSTANCE;
+    private DNSToSwitchMapping dnsResolver;
+    private HashedWheelTimer requestTimer;
+    private FeatureProvider featureProvider;
+
+    public BookKeeperBuilderImpl(ClientConfiguration conf) {
+        this.conf = conf;
+    }
+
+    @Override
+    public BookKeeperBuilder with(Object component) {
+        Preconditions.checkNotNull(component);
+        if (component instanceof ZooKeeper) {
+            this.zk = (ZooKeeper) component;
+        } else if (component instanceof EventLoopGroup) {
+            this.eventLoopGroup = (EventLoopGroup) component;
+        } else if (component instanceof StatsLogger) {
+            this.statsLogger = (StatsLogger) component;
+        } else if (component instanceof DNSToSwitchMapping) {
+            this.dnsResolver = (DNSToSwitchMapping) component;
+        } else if (component instanceof HashedWheelTimer) {
+            this.requestTimer = (HashedWheelTimer) component;
+        } else if (component instanceof FeatureProvider) {
+            this.featureProvider = (FeatureProvider) component;
+        } else {
+            throw new IllegalArgumentException("Component of type " + component.getClass().getName()
+                    + " is not configurable");
+        }
+        return this;
+    }
+
+    @Override
+    public BookKeeper build() throws BKException, InterruptedException {
+        try {
+            return new org.apache.bookkeeper.client.BookKeeper(conf, zk, eventLoopGroup,
```

Comment:
again, why fully qualified path is used here? I saw you used full qualified class path in a lot of places. can you address them?

Reply:
because we have to "BookKeeper" class, the "legacy" o.a.b.client.BookKeeper impl and o.a.b.client.api.BookKeeper interface. For one of the two classes I need to use the FQCN